jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -707,12 +709,12 @@ private void updateKeys()
             // Do update keys for only previously selected keys.
             // This will update only those keys whose selection did not cause an
             // updateKeys update to be submitted.
-            for (SelectionKey key : _keys)
+            _keys.forEach(key ->
```

Comment:
Darn, `HashSet.forEach()` is implemented with an iterator so this doesn't save any allocation.

Reply:
@lorban note that `java.nio.channels.Selector` has new APIs, in particular `select(Consumer<SelectionKey>)`, which may help in not allocating the iterator.
I always wanted to look at this to see if we can improve, but never had time.

BTW, I thought these `Iterator` instances could be easily inlined.

Reply:
@sbordet I don't see how we can use `select(Consumer<SelectionKey>)` as we use the allocated iterator as a cursor, which each processing thread taking the next key and then handling it all the way into the application (if we have reserved threads).
With an iterator, we can have another producing thread take over the iteration, but we could not do so with the new API.

So I think the cost of that iterator allocation is some thing we have to pay for our current scheduling approach... which we should always review to check it is still good.

Reply:
@gregw we checked the two select implementations and they both allocate objects that escape the scope of the call, so they cannot be stack-allocated. It's unavoidable: all NIO select variants do allocate so there's no point in modifying this code.